agent/xds/server.go

```diff
@@ -0,0 +1,418 @@
+		EndpointType: &xDSType{
+			typeURL:   EndpointType,
+			resources: endpointsFromSnapshot,
+			stream:    stream,
+		},
+		ClusterType: &xDSType{
+			typeURL:   ClusterType,
+			resources: clustersFromSnapshot,
+			stream:    stream,
+		},
+		RouteType: &xDSType{
+			typeURL:   RouteType,
+			resources: routesFromSnapshot,
+			stream:    stream,
+		},
+		ListenerType: &xDSType{
+			typeURL:   ListenerType,
+			resources: listenersFromSnapshot,
+			stream:    stream,
+		},
+	}
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			// TODO: Not sure if we should error here, this presumably happens during
+			// a normal shutdown where client goes away too.
+			return nil
+		case req, ok = <-reqCh:
+			if !ok {
+				return nil
+			}
+			if req.TypeUrl == "" {
+				return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+			}
+			if handler, ok := handlers[req.TypeUrl]; ok {
+				handler.Recv(req)
+			}
+		case cfgSnap = <-stateCh:
+			// We got a new config, update the version counter
+			configVersion++
+		}
+
+		// Trigger state machine
+		switch state {
+		case stateInit:
+			if req == nil {
+				// This can't happen (tm) since stateCh is nil until after the first req
+				// is recieved but lets not panic about it.
+				continue
+			}
+			// Start authentication process, we need the proxyID
+			proxyID = req.Node.Id
+
+			// Start watching config for that proxy
+			stateCh, watchCancel = s.cfgMgr.Watch(proxyID)
+			defer watchCancel()
+
+			// Now wait for the config so we can check ACL
+			state = statePendingAuth
+		case statePendingAuth:
+			if cfgSnap == nil {
+				// Nothing we can do until we get the initial config
+				continue
+			}
+			// Got config, try to authenticate
+			token := tokenFromStream(stream)
+			rule, err := s.resolveToken(token)
+			if err != nil {
+				return err
+			}
+			if rule != nil && !rule.ServiceWrite(cfgSnap.Proxy.DestinationServiceName, nil) {
+				return status.Errorf(codes.PermissionDenied, "permission denied")
+			}
+			// Authed OK!
+			state = stateRunning
+
+			// Lets actually process the config we just got or we'll mis responding
+			fallthrough
+		case stateRunning:
+			// See if any handlers need to have the current (possibly new) config
+			// sent. Note the order here is actually significant so we can't just
+			// range the map which has no determined order. It's important because:
+			//
+			//  1. Envoy needs to see a consistent snapshot to avoid potentially
+			//     dropping or misrouting traffic due to inconsitencies. This is the
+			//     main win of ADS after all - we get to control this order.
+			//  2. Non-determinsic order of complex protobug responses which are
+			//     compared for non-exact JSON equivalence makes the tests uber-messy
+			//     to handle
+			for _, typeURL := range []string{ClusterType, EndpointType, RouteType, ListenerType} {
+				handler := handlers[typeURL]
+				err := handler.SendIfNew(cfgSnap, configVersion, &nonce)
```

Comment:
trivial: you can collapse these two lines into into: `if err := ...; err != nil {`

Reply:
I did this although it puts the line 1 char over our typical style guideline... 😄